kvserver: `TestReplicaProbeRequest` is flaky

Reproduces fairly readily under stress on `master`. I've seen two failure modes:

```
=== RUN   TestReplicaProbeRequest
    test_log_scope.go:79: test logs captured to: /home/erik/.cache/bazel/_bazel_erik/58b8816bb117734f0a0eaaef7fcef95b/sandbox/linux-sandbox/677/execroot/com_github_cockroachdb_cockroach/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaProbeRequest1952254550
    test_log_scope.go:80: use -show-logs to present logs inline
    replica_probe_test.go:108: condition failed to evaluate within 45s: saw only 5 probes on s3
        goroutine 83 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x5b779b8, 0xc0000fcd00}, 0x0, 0x1290420)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x5f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x5b779b8, 0xc0000fcd00}, 0x0)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x4a
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaProbeRequest(0xc0000fcd00)
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_probe_test.go:108 +0x69d
        testing.tRunner(0xc0000fcd00, 0x4862238)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
    panic.go:642: -- test log scope end --
test logs left over in: /home/erik/.cache/bazel/_bazel_erik/58b8816bb117734f0a0eaaef7fcef95b/sandbox/linux-sandbox/677/execroot/com_github_cockroachdb_cockroach/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaProbeRequest1952254550
--- FAIL: TestReplicaProbeRequest (46.45s)
```

and

```
=== RUN   TestReplicaProbeRequest
    test_log_scope.go:79: test logs captured to: /home/erik/.cache/bazel/_bazel_erik/58b8816bb117734f0a0eaaef7fcef95b/sandbox/linux-sandbox/728/execroot/com_github_cockroachdb_cockroach/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaProbeRequest1647998711
    test_log_scope.go:80: use -show-logs to present logs inline
    replica_probe_test.go:98: 
        	Error Trace:	replica_probe_test.go:98
        	Error:      	Received unexpected error:
        	            	result is ambiguous: unable to determine whether command was applied via snapshot
        	            	(1) result is ambiguous: unable to determine whether command was applied via snapshot
        	            	Wraps: (2)
        	            	  | (opaque error wrapper)
        	            	  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  | reportable 0:
        	            	  |
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).refreshProposalsLocked
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1286
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1047
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processRaftSnapshotRequest.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:457
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).withReplicaForRequest
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:336
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processRaftSnapshotRequest
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:395
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).receiveSnapshot
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_snapshot.go:788
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).HandleSnapshot.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:211
        	            	  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:344
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).HandleSnapshot
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:208
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).RaftSnapshot.func1.1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/raft_transport.go:525
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).RaftSnapshot.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/raft_transport.go:526
        	            	  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1581
        	            	Wraps: (3) unable to determine whether command was applied via snapshot
        	            	Error types: (1) *roachpb.AmbiguousResultError (2) *errbase.opaqueWrapper (3) *errutil.leafError
        	Test:       	TestReplicaProbeRequest
    panic.go:642: -- test log scope end --
test logs left over in: /home/erik/.cache/bazel/_bazel_erik/58b8816bb117734f0a0eaaef7fcef95b/sandbox/linux-sandbox/728/execroot/com_github_cockroachdb_cockroach/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaProbeRequest1647998711
--- FAIL: TestReplicaProbeRequest (2.37s)
```

Jira issue: CRDB-15197

--- a/pkg/kv/kvserver/replica_probe_test.go
+++ b/pkg/kv/kvserver/replica_probe_test.go
@@ -58,6 +58,41 @@ func TestReplicaProbeRequest(t *testing.T) {
 		}
 		return 0, args.ForcedError
 	}
+	// NB: At one point we tried to verify that the probe the tests sends below
+	// is seen in command application on *all* replicas.
+	//
+	// This was flaky[^1] due to snapshot application (g1 and g2 are the
+	// intertwined operations):
+	//
+	// - g1    | `tc.AddXOrFatal` begins
+	// - g1    | it calls into ChangeReplicas
+	// - g1    | Replica added to descriptor
+	//      g2 | leader sends first MsgApp
+	// -    g2 | follower gets MsgApp, rejects it, leader enqueues in Raft snapshot queue
+	// - g1    | block raft snapshots to follower
+	// - g1    | send INITIAL snapshot
+	// - g1    | unblock raft snapshots to follower
+	// - g1    | AddVoterOrWhatever returns to the test
+	// - g1    | test sends probe
+	// -    g2 | raft log queue processes the queued Replica (followers's
+	//           MsgAppResp following INITIAL snapshot not having reached leader yet)[^1]
+	// -    g2 | it sends another snap which includes the probe
+	// -    g2 | follower applies the probe via the snap.
+	// - g1    | times out waiting for the interceptor to see the probe on follower.
+	//
+	// We attempted to disable the raft snapshot queue. This caused deadlocks: if raft
+	// (erroneously or not) requests a snapshot, it will not replicate to the follower
+	// until it has been caught up. But the quorum size here is two, so this would cause
+	// a hanging test. We tried to work around this by marking snapshots as failed when
+	// the raft snapshot queue is disabled, but this then threw off other tests[^2] that
+	// wanted to observe this state.
+	//
+	// So we just check that the probe applies on at least one Replica, which
+	// always has to be true.
+	//
+	// [^1]: https://github.com/cockroachdb/cockroach/pull/92380
+	// [^2]: TestMigrateWithInflightSnapshot, TestSnapshotsToDrainingNodes, TestRaftSnapshotQueueSeesLearner
+
 	args.ServerArgs.Knobs.Store = &kvserver.StoreTestingKnobs{
 		// We set an ApplyFilter even though the probe should never
 		// show up there (since it always catches a forced error),
@@ -111,7 +146,10 @@ func TestReplicaProbeRequest(t *testing.T) {
 		if exp, act := len(seen.m), len(tc.Servers); exp != act {
 			return errors.Errorf("waiting for stores to apply command: %d/%d", act, exp)
 		}
-		n := 2 * len(tc.Servers) // sent two probes per server
+		// We'll usually see 2 * len(tc.Servers) probes since we sent two probes, but see
+		// the comment about errant snapshots above. We just want this test to be reliable
+		// so expect at least one probe in command application.
+		n := 1
 		for storeID, count := range seen.m {
 			if count < n {
 				return errors.Errorf("saw only %d probes on s%d", count, storeID)

